presto-raptorx/src/main/java/com/facebook/presto/raptorx/chunkstore/ChunkStoreCleaner.java

```diff
@@ -0,0 +1,229 @@
+    public ChunkStoreCleaner(ChunkStore chunkStore, Database database, ChunkStoreCleanerConfig config)
+    {
+        this(chunkStore, database, config.getInterval(), config.getThreads());
+    }
+
+    public ChunkStoreCleaner(ChunkStore chunkStore, Database database, Duration interval, int threads)
+    {
+        this.chunkStore = requireNonNull(chunkStore, "chunkStore is null");
+        this.interval = requireNonNull(interval, "interval is null");
+        this.coreExecutor = newCachedThreadPool(daemonThreadsNamed("chunk-store-cleaner-%s"));
+        this.executor = new BoundedExecutor(coreExecutor, threads);
+        this.dao = createJdbi(database.getMasterConnection()).onDemand(CleanerDao.class);
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!started.getAndSet(true)) {
+            scheduler.scheduleWithFixedDelay(this::run, 0, interval.toMillis(), MILLISECONDS);
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        scheduler.shutdownNow();
+        coreExecutor.shutdownNow();
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getJobErrors()
+    {
+        return jobErrors;
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getDeleteFailures()
+    {
+        return deleteFailures;
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getDeleteSuccesses()
+    {
+        return deleteSuccesses;
+    }
+
+    private void run()
+    {
+        try {
+            cleanChunkStore();
+        }
+        catch (Throwable t) {
+            log.error(t, "Error cleaning chunk store");
+            jobErrors.update(1);
+        }
+    }
+
+    public synchronized void cleanChunkStore()
+    {
+        Set<Long> processing = newConcurrentHashSet();
+        BlockingQueue<Long> completed = new LinkedBlockingQueue<>();
+        boolean fill = true;
+
+        while (!Thread.currentThread().isInterrupted()) {
+            // get a new batch if any completed and we are under the batch size
+            Set<Long> chunkIds = ImmutableSet.of();
+            if (fill && (processing.size() < CHUNK_BATCH_SIZE)) {
+                chunkIds = dao.getCleanableChunksBatch(System.currentTimeMillis(), CHUNK_BATCH_SIZE);
+                fill = false;
+            }
+            if (chunkIds.isEmpty() && processing.isEmpty()) {
+                break;
+            }
+
+            // skip any that are already processing and mark remaining as processing
+            chunkIds = ImmutableSet.copyOf(difference(chunkIds, processing));
+            processing.addAll(chunkIds);
+
+            // execute deletes
+            for (long chunkId : chunkIds) {
+                runAsync(() -> chunkStore.deleteChunk(chunkId), executor)
+                        .thenAccept(v -> completed.add(chunkId))
+                        .whenComplete((v, e) -> {
+                            if (e != null) {
+                                log.error(e, "Error deleting chunk from chunk store: %s", chunkId);
```

Comment:
highker on Jan 30  Member
There might be a race here:

the root thread just finished chunkIds = dao.getCleanableChunksBatch(System.currentTimeMillis(), CHUNK_BATCH_SIZE); and suppose chunkIdsis empty.
a forked thread finished processing.remove(chunkId); and processing might be empty.
the root thread checks if (chunkIds.isEmpty() && processing.isEmpty()) and returns.
If cleanChunkStore is only "try our best" to clean chun store, this might be ok if the next scheduled run the failed chunkId can still be picked up by dao.getCleanableChunksBatch.

  @zhaotao1986
zhaotao1986 on Feb 14  Member
Please refer to the quip about the full analysis of ChunkStoreCleaner